test/test_ion_text.cpp

```diff
@@ -295,21 +306,35 @@ TEST(IonTextSymbol, WriterWriteAllValuesPreservesSymbolZero) {
     const BYTE *ion_binary = (BYTE *)"\xDE\x90\x8A\xE3\x81\x8A\x70\x80\xE4\x81\x80\x71\x0A\x80\xE3\x81\x8A\x70";
     const SIZE ion_binary_size = 18;
 
-    BYTE *result = NULL;
-    SIZE result_len;
+    // Storage and length to receive text encoding.
+    BYTE *as_text = NULL;
+    SIZE as_text_len = 0;
 
-    ion_test_write_all_values_from_text(ion_text, &result, &result_len, FALSE);
-    assertStringsEqual(ion_text, (char *)result, result_len);
+    ion_test_write_all_values_from_text(ion_text, &as_text, &as_text_len, FALSE);
+    assertStringsEqual(ion_text, (char *)as_text, as_text_len);
+    free(as_text); as_text = NULL;
 
-    ion_test_write_all_values_from_text(ion_text, &result, &result_len, TRUE);
-    assertBytesEqual((const char *)ion_binary, ion_binary_size, result + result_len - ion_binary_size, ion_binary_size);
+    // Storage and length to receive binary encoding.
+    BYTE *as_binary = NULL;
+    SIZE as_binary_len = 0;
 
-    ion_test_write_all_values_from_binary(result, result_len, &result, &result_len, FALSE);
```

Comment:
With the way `ion_test_write_all_values_from_binary` is written, the destination `&result` is expected to be a pointer to a pointer that can receive the address of the buffer that the function will allocate. So re-using `result` here results in the previous result buffer being leaked, and result containing a new address. This occurs again below. I added more pointers & lengths to track and free the individual buffers.